docs/proposals/2019-02-08 - Fallback Content Elements.md

```diff
@@ -0,0 +1,84 @@
+# Alternatives to Raw HTML
+
+# Problem
+
+The content element `raw_html` is often used to inject Javascript into an ANS document which makes it difficult for downstream applications to render natively. Even if the application tries to render it as a embeded browser, there is no guarentee that it will render correctly in all applications. And some applications (ex Apple TV or $50 Kindle) may not have the input or processing capabilities needed to effectively render the experience.
+
+# Proposal
+
+Add an `alternative` trait to the `raw_html` content element, that contains an array of one content elements that can notionally replace the `raw_html` block. The content in the array can only be of type `text`, `image`, `gallery` or `video`. This trait is required unless the `rendering_guides.preferred_method` is set to `website` only. In all other cases, when `upvert`ing this trait should be filled this with a `text` element that has a stock `This content is not available on this platform` message (or similar).
+
+## Alternatives
+
+Each `raw_html` element should have an `alternative` and at least a `basic` array with non-`raw_html` content. An example of the default behaviour:
+
+```
+{
+  "type": "raw_html",
+  "content": "<script src=\"https://www.washingtonpost.com/stat/cool.js\"></script>",
+  "alternative": [
+    {
+      "type": "text",
+      "content": "<b>This content is unavailable on this platform</b>"
+    }
+  ]
+}
+```
+
+# Example 1
+
+The graphics team has created a JS rendered graphic that scales to browser viewport size. They also generate static `.png`s as a fallback for mobile and tablet.
+
+```
+{
+  "type": "raw_html",
+  "content": "<div class=\"graphics\"><!-- Graphics HTML and JS --></div>",
+  "alternative": [
+    {
+      "type": "image",
+      "url": "https://www.coolpublishing.com/graphics/alt-cool-1.png"
+    }
+  ]
+}
+```
+
+# Example 2 (future implementation of multiple elements)
+
+The social team is trying to build awareness of the Twitch channel so they are embedding the Twitch stream where normally we would use the Goldfish live video. They provide the normal video as an alternative as well as a note to encourage users to open the twitch channel.
+
+```
+{
+  "type": "raw_html",
+  "content": "<iframe src=\"https://player.twitch.tv/?channel=coolpublishing\" frameborder=\"0\" allowfullscreen=\"true\" scrolling=\"no\" height=\"378\" width=\"620\"></iframe><a href=\"https://www.twitch.tv/coolpublishing?tt_content=text_link&tt_medium=live_embed\" style=\"padding:2px 0px 4px; display:block; width:345px; font-weight:normal; font-size:10px; text-decoration:underline;\">Watch live video from coolpublishing on www.twitch.tv</a>",
+  "alternative": [
```

Comment:
As you point out, determining if the `raw_html` rendering failed is difficult, but this proposal does not try to address that case. It is intended to be a fallback for rendering platforms that otherwise ignore the `raw_html` element or reject a story because it has that element. The rendering platform should make the determination itself depending on if it supports open-ended html. Many of the rendering platforms other than the website fall into this category (mobile apps, facebook instant, apple news, google news, etc) and `raw_html` can make the article confusing if important content is contained in it and it's not shown. This provides a mechanism to allow for a passible alternative, or to inform the reader if there is something not shown when `raw_html` is not supported.